### PR TITLE
Allow sending data frames with padding

### PR DESCRIPTION
**Pull request by apolcyn**
This allows sending padding along with a data frame.

This defaults `pad_length` to `None`, indicating that padding flag should be unset - wanted to allow setting zero length padding with a zero-valued `pad_length`.

### CODE REVIEW

File: h2/connection.py [hunk 1]
```diff
@@ -858,25 +858,35 @@ def send_data(self, stream_id, data, end_stream=False):
         :type data: ``bytes``
         :param end_stream: (optional) Whether this is the last data to be sent
             on the stream. Defaults to ``False``.
+        :param pad_length: (optional) Length of the padding to apply to the
+            data frame. Defaults to ``None`` for no use of padding. Note that
+            a value of ``0`` results in padding of length ``0``
+            (with the "padding" flag set on the frame).
```
Comment (alexwlchan) [inline h2/connection.py]: Please add a `versionadded: 2.6.0` directive to this parameter ([example](https://github.com/apolcyn/hyper-h2/blob/4322904ac68cce2f5271034199cf400a82e289de/h2/connection.py#L294)), so the documentation will reflect when it became available.

File: h2/connection.py [hunk 2]
```diff
@@ -858,25 +858,35 @@ def send_data(self, stream_id, data, end_stream=False):
...
         :type end_stream: ``bool``
         :returns: Nothing
         """
-        if len(data) > self.local_flow_control_window(stream_id):
+        frame_size = len(data)
+        if pad_length is not None:
+            frame_size += pad_length + 1
```
Comment (alexwlchan) [inline h2/connection.py]: Maybe it’s too early in the morning – why do we need the `+1`? (if it’s not obvious, worth a comment?)

File: h2/stream.py [hunk 3]
```diff
@@ -850,12 +850,17 @@ def send_data(self, data, end_stream=False):
...
         df.data = data
         if end_stream:
             self.state_machine.process_input(StreamInputs.SEND_END_STREAM)
             df.flags.add('END_STREAM')
+        if pad_length is not None:
+            df.flags.add('PADDED')
+            df.pad_length = pad_length
 
-        self.outbound_flow_control_window -= len(data)
+        # Subtract flow_controll_length to account for possible padding
```
Comment (alexwlchan) [inline h2/stream.py]: /s/flow_controll_length/flow_controlled_length

File: test/test_flow_control_window.py [hunk 4]
```diff
@@ -52,6 +52,28 @@ def test_flow_control_decreases_with_sent_data(self):
...
+    def test_flow_control_decreases_with_sent_data_with_padding(self):
+        """
+        When data is sent on a stream, the flow control window should drop.
+        """
+        c = h2.connection.H2Connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.send_data(1, b'some data', pad_length=5)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 6
+        assert (c.local_flow_control_window(1) == remaining_length)
```
Comment (alexwlchan) [inline test/test_flow_control_window.py]: Parentheses are unnecessary.

File: test/test_flow_control_window.py [hunk 5]
```diff
@@ -52,6 +52,28 @@ def test_flow_control_decreases_with_sent_data(self):
...
+    def test_flow_control_decreases_with_sent_data_zero_length_padding(self):
+        """
+        When data is sent on a stream, the flow control window should drop.
+        """
+        c = h2.connection.H2Connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.send_data(1, b'some data', pad_length=0)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 1
+        assert (c.local_flow_control_window(1) == remaining_length)
```
Comment (alexwlchan) [inline test/test_flow_control_window.py]: Ditto above.

File: test/test_basic_logic.py [hunk 6]
```diff
@@ -113,6 +113,40 @@ def test_sending_data(self):
...
+        c.send_headers(1, self.example_request_headers)
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
+        events = c.send_data(1, b'some data', pad_length=0)
+        assert not events
+        assert c.data_to_send() == (
+            b'\x00\x00\x0a\x00\x08\x00\x00\x00\x01'
+            b'\x00some data'
+        )
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: I’d be keen on adding an `assert` to these tests that the "PADDED" flag is present in both cases. That might already be encoded in `data_to_send()`, but if a human could see that without unparsing that data. :-)

Comment (alexwlchan) [inline test/test_basic_logic.py]: Also, I think we should have a test for the fallback behaviour: if you pass `pad_length`, your frame _doesn’t_ have the "PADDED" flag set.

File: h2/connection.py (hunk 2 above)
Comment (Lukasa) [inline h2/connection.py]: So, I know why we need the `+1`, but it's still worth a comment expressly because it's unclear unless you've read the RFC fairly closely.

(@alexwlchan an FYI, the reason is that when you add the PADDING flag you don't just add the padding bytes, you also add an extra byte that indicates how many padding bytes there are. That byte is also part of the flow controlled length.)

File: h2/connection.py [hunk 7]
```diff
@@ -858,25 +858,35 @@ def send_data(self, stream_id, data, end_stream=False):
...
             raise FrameTooLargeError(
                 "Cannot send frame size %d, max frame size is %d" %
-                (len(data), self.max_outbound_frame_size)
+                (frame_size, self.max_outbound_frame_size)
             )
 
         self.state_machine.process_input(ConnectionInputs.SEND_DATA)
-        frames = self.streams[stream_id].send_data(data, end_stream)
+        frames = self.streams[stream_id].send_data(
+            data, end_stream, pad_length=pad_length)
```
Comment (Lukasa) [inline h2/connection.py]: This is a style thing for the project, but can the closing parenthesis be moved to a line of its own? See what is done for the `FrameTooLargeError` above.

File: h2/stream.py [hunk 8]
```diff
@@ -850,12 +850,17 @@ def send_data(self, data, end_stream=False):
         self.state_machine.process_input(StreamInputs.SEND_DATA)
 
         df = DataFrame(self.stream_id)
+
```
Comment (Lukasa) [inline h2/stream.py]: We don't really need this whitespace line.

File: h2/connection.py (hunk 2 above)
Comment (Lukasa) [inline h2/connection.py]: Can we please validate the input? That is, we allow `pad_length` to be `None` or an integer that is `>= 0`. I'd like us to actually check that and to raise a `ValueError` if it's wrong.

Note that you can go ahead and raise the `ValueError` inside this `if` block if you want, which saves checking twice.

File: test/test_flow_control_window.py [hunk 9]
```diff
@@ -52,6 +52,28 @@ def test_flow_control_decreases_with_sent_data(self):
         remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data')
         assert (c.local_flow_control_window(1) == remaining_length)
 
+    def test_flow_control_decreases_with_sent_data_with_padding(self):
+        """
+        When data is sent on a stream, the flow control window should drop.
```
Comment (Lukasa) [inline test/test_flow_control_window.py]: Do you mind updating this docstring to be unique and to explain specifically what it tests? That is, that the flow control window drops by an amount that includes the padding?

File: test/test_flow_control_window.py [hunk 10]
```diff
@@ -52,6 +52,28 @@ def test_flow_control_decreases_with_sent_data(self):
...
+        c = h2.connection.H2Connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.send_data(1, b'some data', pad_length=5)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 6
+        assert (c.local_flow_control_window(1) == remaining_length)
+
+    def test_flow_control_decreases_with_sent_data_zero_length_padding(self):
+        """
+        When data is sent on a stream, the flow control window should drop.
```
Comment (Lukasa) [inline test/test_flow_control_window.py]: Same note here: this docstring needs to be updated to note that we're testing that with zero-length padding the flow control window should drop by one byte more than the data we sent.

File: h2/connection.py (hunk 2 above)
Comment (SethMichaelLarson) [inline h2/connection.py]: Also can't be longer than 255 if I remember?

Comment (Lukasa) [inline h2/connection.py]: Ah, good spot, also true.

**Comment by apolcyn** [comment]
tried to address comments in latest 3 commits. 

@alexwlchan  wanted to reach out on one though :)
> ... adding an assert to these tests that the "PADDED" flag is present ...
A bit stuck, is there a convenience helper for grabbing the `DataFrame` that's to send, or parsing one from the `data_to_send()`?

File: h2/connection.py [hunk 11]
```diff
@@ -859,24 +859,40 @@ def send_data(self, stream_id, data, end_stream=False):
         :param end_stream: (optional) Whether this is the last data to be sent
             on the stream. Defaults to ``False``.
         :type end_stream: ``bool``
+        :param pad_length: (optional) Length of the padding to apply to the
+            data frame. Defaults to ``None`` for no use of padding. Note that
+            a value of ``0`` results in padding of length ``0``
+            (with the "padding" flag set on the frame).
+        :type pad_length: ``int``
+            .. versionadded:: 2.6.0
```
Comment (alexwlchan) [inline h2/connection.py]: If you build this the docs locally, the directive doesn’t show up correctly. I confess I don’t fully understand how Sphinx directives work, but this block of RST seems to do the trick:

```rst
        :param pad_length: (optional) Length of the padding to apply to the
            data frame. Defaults to ``None`` for no use of padding. Note that
            a value of ``0`` results in padding of length ``0``
            (with the "padding" flag set on the frame).

            .. versionadded:: 2.6.0

        :type pad_length: ``int``
```

File: h2/connection.py [hunk 12]
```diff
@@ -859,24 +859,40 @@ def send_data(self, stream_id, data, end_stream=False):
...
         :returns: Nothing
         """
-        if len(data) > self.local_flow_control_window(stream_id):
+        frame_size = len(data)
+        if pad_length is not None:
+            # Account for padding bytes and the 1-byte padding length field.
+            if pad_length < 0 or pad_length > 255:
+                raise ValueError("pad length must be within range: [0, 255]")
```
Comment (alexwlchan) [inline h2/connection.py]: What if I pass in a non-integer padding length?

Also, tweak the error message to use the exact name of the parameter (with the underscore).

File: h2/connection.py [hunk 13]
```diff
@@ -859,24 +859,40 @@ def send_data(self, stream_id, data, end_stream=False):
...
+            frame_size += pad_length + 1
```
Comment (alexwlchan) [inline h2/connection.py]: Still missing a comment about why we have the extra `+1`.

File: test/test_basic_logic.py [hunk 14]
```diff
@@ -113,6 +113,57 @@ def test_sending_data(self):
...
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.clear_outbound_data_buffer()
+        # 255 is valid
+        c.send_data(1, b'some data', pad_length=255)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=-1)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=256)
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: We’ll want some extra tests here that we throw a `TypeError` if passing a non-integer padding length.

File: h2/connection.py (hunk 13 above)
Comment (alexwlchan) [inline h2/connection.py]: (No, putting it into the tests doesn’t count. 😉 I might look at the tests if I’ve broken something, but it doesn’t help somebody who’s just casually reading the code.)

**Comment by Lukasa** [comment]
> is there a convenience helper for grabbing the DataFrame that's to send, or parsing one from the data_to_send()?

Yes, there is. Just import the frame buffer from the codebase. You can see this done in `test_basic_logic` `test_oversize_headers`.

File: h2/connection.py (hunk 12 above)
Comment (Lukasa) [inline h2/connection.py]: I'm inclined to want to leave that off: on Python 3 they'll mostly raise type errors, and on Python 2 it'll fail lower down the stack.

**Comment by apolcyn** [comment]
latest two commits try to address recent comments again.

added in some simple type checks on the pad_length param too.

thanks for comments :) I think it's ready for another look

File: test/test_basic_logic.py [hunk 15]
```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
...
+        assert len(frames) == 3
+        settings_frame = frames[0]
+        header_frame = frames[1]
+        data_frame = frames[2]
+        assert isinstance(settings_frame, hyperframe.frame.SettingsFrame)
+        assert isinstance(header_frame, hyperframe.frame.HeadersFrame)
+        assert isinstance(data_frame, hyperframe.frame.DataFrame)
+
+        assert 'PADDED' in data_frame.flags
+
```
Comment (Lukasa) [inline test/test_basic_logic.py]: So there is a bunch of duplicate code around checking the output of the stream and whether there is a padding flag, and if so how much. It might be nice to factor that out to a helper function.

File: test/test_basic_logic.py [hunk 16]
```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
...
+        frames = list(buffer)
+        assert len(frames) == 3
+        settings_frame = frames[0]
+        header_frame = frames[1]
+        data_frame = frames[2]
+        assert isinstance(settings_frame, hyperframe.frame.SettingsFrame)
+        assert isinstance(header_frame, hyperframe.frame.HeadersFrame)
+        assert isinstance(data_frame, hyperframe.frame.DataFrame)
+
+        assert 'PADDED' in data_frame.flags
```
Comment (Lukasa) [inline test/test_basic_logic.py]: Let's also check the amount of padding and flow controlled length, just to cover all our bases?

File: test/test_basic_logic.py [hunk 17]
```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
...
+
+    def test_sending_data_with_invalid_padding_length(self):
+        """
+        Pad_lengths < 0 or > 255 are be checked for and rejected.
```
Comment (Lukasa) [inline test/test_basic_logic.py]: You can remove `be` from this sentence.

File: test/test_basic_logic.py [hunk 18]
```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
...
+        with pytest.raises(TypeError):
+            c.send_data(1, b'some data', pad_length='')
+        with pytest.raises(TypeError):
+            c.send_data(1, b'some data', pad_length='10')
+        with pytest.raises(TypeError):
+            c.send_data(1, b'some data', pad_length={})
+        with pytest.raises(TypeError):
+            c.send_data(1, b'some data', pad_length=[1, 2, 3])
+        with pytest.raises(TypeError):
+            c.send_data(1, b'some data', pad_length=[])
```
Comment (Lukasa) [inline test/test_basic_logic.py]: This test would be nicer if rewritten using`pytest.mark.parametrize`.

File: test/test_basic_logic.py [hunk 19]
```diff
@@ -105,14 +105,142 @@ def test_sending_data(self):
...
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.clear_outbound_data_buffer()
+        # 255 is valid
+        c.send_data(1, b'some data', pad_length=255)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=-1)
+        with pytest.raises(ValueError):
+            c.send_data(1, b'some data', pad_length=256)
```
Comment (Lukasa) [inline test/test_basic_logic.py]: This test would be nicer if written with `pytest.mark.parametrize`.

File: test/test_flow_control_window.py [hunk 20]
```diff
@@ -52,6 +52,32 @@ def test_flow_control_decreases_with_sent_data(self):
         remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data')
         assert (c.local_flow_control_window(1) == remaining_length)
 
+    def test_flow_control_decreases_with_sent_data_with_padding(self):
+        """
+        When padded data is sent on a stream, the flow control window should
+        drop by the length of the padding plus 1 for the 1-byte padding length
+        field.
```
Comment (Lukasa) [inline test/test_flow_control_window.py]: This is a minor nit, but we try to write these docstrings as imperative: that is, we don't write that X *should* happen, we write that it *does* happen. For example, this would be:

> When padded data is sent on a stream, the flow control window drops by the length of the padding plus 1 for the 1-byte padding length field.

File: test/test_flow_control_window.py [hunk 21]
```diff
@@ -52,6 +52,32 @@ def test_flow_control_decreases_with_sent_data(self):
...
+
+        c.send_data(1, b'some data', pad_length=5)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 6
+        assert c.local_flow_control_window(1) == remaining_length
+
+    def test_flow_control_decreases_with_sent_data_zero_length_padding(self):
+        """
+        When zero-length-padded data is sent on a stream, the flow control
+        window should drop by the 1 for the 1-byte padding length field, which
+        is still present though zero-valued.
```
Comment (Lukasa) [inline test/test_flow_control_window.py]: Again, this should be imperative.

File: test/test_basic_logic.py (hunk 16 above)
Comment (apolcyn) [inline test/test_basic_logic.py]: err, while adding extra checks here, I think I encountered problem!

Currently it looks to me like the `hyperframe.DataFrame` class has an issue in the way it computes the flow controlled length. I think the fix is small, created https://github.com/python-hyper/hyperframe/pull/80 for it

**Comment by apolcyn** [comment]
thanks for comments again :) latest updates tried to address.

extra checks seem to have pointed out an issue though, not sure what we should do about this one in the meantime since the "zero valued pad length" test will fail for the time being (fix in https://github.com/python-hyper/hyperframe/pull/80)

**Comment by Lukasa** [comment]
I propose we shelve this patch until a new version of hyperframe is ready.

**Comment by apolcyn** [comment]
Shelving this until the hyperframe update is published sounds good to me, thanks for the review on that!

**Comment by Lukasa** [comment]
Ok, hyperframe v4.0.2 has been shipped and is ready to go. This can be unblocked now. =D

**Comment by apolcyn** [comment]
great, much thanks! 

So I think the last updates should address the comments again so far. (Added a new helper function in `test/helpers.py` - but not completely sure where it should go..)

For me, tests passed locally after clearing out my `.tox` and forcing the update to the new `hyperframe` package. I'm not sure what needs to be done here, but if doable I believe that a similar sort of test re-run would work here on travis.

File: test/helpers.py [hunk 22]
```diff
@@ -167,3 +168,29 @@ def build_alt_svc_frame(self, stream_id, origin, field):
...
+        assert(
+            (data_frame.flow_controlled_length ==
+             expected_flow_controlled_length)
+        )
+        if expect_padded_flag:
+            assert 'PADDED' in data_frame.flags
+        else:
+            assert 'PADDED' not in data_frame.flags
+
+        assert data_frame.pad_length == expected_data_frame_pad_length
```
Comment (Lukasa) [inline test/helpers.py]: Given that this is only used in one test location, I think it'd be better to make it a helper function in that file rather than place it in a central location.

File: test/test_basic_logic.py [hunk 23]
```diff
@@ -105,14 +105,118 @@ def test_sending_data(self):
...
-            c.data_to_send() == b'\x00\x00\t\x00\x00\x00\x00\x00\x01some data'
+            data_to_send == b'\x00\x00\t\x00\x00\x00\x00\x00\x01some data'
+        )
+
+        buffer.add_data(data_to_send)
+        helpers.sanity_check_settings_header_data(
+            frames=list(buffer),
+            expected_flow_controlled_length=len(b'some data'),
+            expect_padded_flag=False,
+            expected_data_frame_pad_length=0)
```
Comment (Lukasa) [inline test/test_basic_logic.py]: Can you drop this closing parenthesis to the next line please?

File: test/test_basic_logic.py [hunk 24]
```diff
@@ -105,14 +105,118 @@ def test_sending_data(self):
...
+            b'\x00\x00\x0f\x00\x08\x00\x00\x00\x01'
+            b'\x05some data\x00\x00\x00\x00\x00'
         )
 
+        buffer.add_data(data_to_send)
+        helpers.sanity_check_settings_header_data(
+            frames=list(buffer),
+            expected_flow_controlled_length=len(b'some data') + 1 + 5,
+            expect_padded_flag=True,
+            expected_data_frame_pad_length=5)
```
Comment (Lukasa) [inline test/test_basic_logic.py]: Same closing parenthesis note here.

File: test/test_basic_logic.py [hunk 25]
```diff
@@ -105,14 +105,118 @@ def test_sending_data(self):
...
+            b'\x00\x00\x0a\x00\x08\x00\x00\x00\x01'
+            b'\x00some data'
+        )
+
+        buffer.add_data(data_to_send)
+        helpers.sanity_check_settings_header_data(
+            frames=list(buffer),
+            expected_flow_controlled_length=len(b'some data') + 1,
+            expect_padded_flag=True,
+            expected_data_frame_pad_length=0)
```
Comment (Lukasa) [inline test/test_basic_logic.py]: Same closing parenthesis note here.

File: test/test_basic_logic.py [hunk 26]
```diff
@@ -105,14 +105,118 @@ def test_sending_data(self):
...
+        ([], TypeError),
+    ])
+    def test_sending_data_with_invalid_padding_length(self,
+                                                      pad_length,
+                                                      expected_error):
+        """
+        Integer valued pad_length values < 0 or > 255 are checked for
+        and rejected.
+        Also ``send_data`` with a ``pad_length`` parameter that is neither
+        ``None`` or an ``integer`` raises a ``TypeError``
```
Comment (Lukasa) [inline test/test_basic_logic.py]: I'd rather have this docstring be one sentence if we can do it.

File: test/test_flow_control_window.py (hunk 20 above)
Comment (Lukasa) [inline test/test_flow_control_window.py]: This still needs to be made imperative.

File: test/test_flow_control_window.py [hunk 27]
```diff
@@ -52,6 +52,32 @@ def test_flow_control_decreases_with_sent_data(self):
...
+        c.send_headers(1, self.example_request_headers)
+
+        c.send_data(1, b'some data', pad_length=5)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 6
+        assert c.local_flow_control_window(1) == remaining_length
+
+    def test_flow_control_decreases_with_sent_data_zero_length_padding(self):
+        """
+        When zero-length-padded data is sent on a stream, the flow control
+        window is drop by the 1 for the 1-byte padding length field (which
```
Comment (Lukasa) [inline test/test_flow_control_window.py]: Nit: we don't want "is drop", we want "drops".

**Comment by apolcyn** [comment]
Latest update should address comments again.

I think the introduced `non-imperative` should be fixed..

File: test/test_basic_logic.py [hunk 28]
```diff
@@ -1576,3 +1683,29 @@ def test_receiving_goaway_frame_with_unknown_error(self, frame_factory):
         assert c.state_machine.state == h2.connection.ConnectionState.CLOSED
 
         assert not c.data_to_send()
+
+
+def sanity_check_settings_header_data(frames,
+                                      expected_flow_controlled_length,
+                                      expect_padded_flag,
+                                      expected_data_frame_pad_length):
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: Let’s get a docstring on this function to explain what it’s checking.

File: test/test_basic_logic.py [hunk 29]
```diff
@@ -1576,3 +1683,29 @@ def test_receiving_goaway_frame_with_unknown_error(self, frame_factory):
...
+        assert len(frames) == 3
+        settings_frame = frames[0]
+        header_frame = frames[1]
+        data_frame = frames[2]
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: Alternative:

```python
settings_frame, header_frame, data_frame = frames
```

File: test/test_flow_control_window.py [hunk 30]
```diff
@@ -52,6 +52,32 @@ def test_flow_control_decreases_with_sent_data(self):
...
+        When zero-length-padded data is sent on a stream, the flow control
+        window drops by the 1 for the 1-byte padding length field (which
+        is still present though zero-valued).
+        """
+        c = h2.connection.H2Connection()
+        c.send_headers(1, self.example_request_headers)
+
+        c.send_data(1, b'some data', pad_length=0)
+        remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - 1
+        assert c.local_flow_control_window(1) == remaining_length
```
Comment (alexwlchan) [inline test/test_flow_control_window.py]: These two tests are very similar. Could we combine them into two tests which take a `padding_length` parameter, and then define:

```python
remaining_length = self.DEFAULT_FLOW_WINDOW - len(b'some data') - padding_length - 1
```

?

File: test/test_basic_logic.py [hunk 31]
```diff
@@ -105,14 +105,121 @@ def test_sending_data(self):
...
+        (255, None),
+        (None, None),
+        (-1, ValueError),
+        (256, ValueError),
+        ('invalid', TypeError),
+        ('', TypeError),
+        ('10', TypeError),
+        ({}, TypeError),
+        (['1', '2', '3'], TypeError),
+        ([], TypeError),
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: Style nit: I’d consider rewriting this to put the error first. That way cases line up a bit more neatly:

```
(None, 0)
(None, 255)
(None, None)
(ValueError, -1)
(ValueError, 256)
(TypeError, 'invalid')
...
```

It’s easy to skim and see where the `ValueError`'s stop and the `TypeError`s start.

Comment (alexwlchan) [inline test/test_basic_logic.py]: Also, let’s throw in a floating point number, which could be integral but is still disallowed.

**Comment by apolcyn** [comment]
latest update tries to address comments from @alexwlchan (in commit b0ca7fc) 

not super confident in the docstring for `sanity_check_settings_header_data` but I think it's ready for another look.

File: test/test_basic_logic.py [hunk 32]
```diff
@@ -1576,3 +1686,34 @@ def test_receiving_goaway_frame_with_unknown_error(self, frame_factory):
...
+
+def sanity_check_settings_header_data(frames,
+                                      expected_flow_controlled_length,
+                                      expect_padded_flag,
+                                      expected_data_frame_pad_length):
+        """
+        ```frames``` is a list of http2 frames in order of types
+        "settings", "header", and "data", and the ``flags``
+        and ```flow_controlled_length``` of the "data frame" match
+        expectations.
```
Comment (Lukasa) [inline test/test_basic_logic.py]: So, while I was writing a docstring for this function I realised that it is just doing some unnecessary work. I think we should change it so that it doesn't bother to receive the first two frames, as it doesn't do any meaningful validation of them, and it limits the scope of the function to only test straight after connection setup.

We can just ignore the first two frames and assume that they're not relevant in the tests.

**Comment by apolcyn** [comment]
@Lukasa  sounds good, latest update just tried to simplify the helper function to only check the data frame.

File: test/test_basic_logic.py [hunk 33]
```diff
@@ -105,14 +105,127 @@ def test_sending_data(self):
...
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        buffer = h2.frame_buffer.FrameBuffer(server=True)
+        buffer.max_frame_size = 65535
+        buffer.add_data(c.data_to_send())
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
```
Comment (Lukasa) [inline test/test_basic_logic.py]: This shouldn't be needed if you've called `data_to_send()`.

File: test/test_basic_logic.py [hunk 34]
```diff
@@ -105,14 +105,127 @@ def test_sending_data(self):
         c.initiate_connection()
         c.send_headers(1, self.example_request_headers)
 
+        buffer = h2.frame_buffer.FrameBuffer(server=True)
+        buffer.max_frame_size = 65535
+        buffer.add_data(c.data_to_send())
+
         # Clear the data, then send some data.
         c.clear_outbound_data_buffer()
```
Comment (Lukasa) [inline test/test_basic_logic.py]: This is no longer needed.

File: test/test_basic_logic.py [hunk 35]
```diff
@@ -105,14 +105,127 @@ def test_sending_data(self):
...
+        c = h2.connection.H2Connection()
+        c.initiate_connection()
+        c.send_headers(1, self.example_request_headers)
+
+        buffer = h2.frame_buffer.FrameBuffer(server=True)
+        buffer.max_frame_size = 65535
+        buffer.add_data(c.data_to_send())
+
+        # Clear the data, then send some data.
+        c.clear_outbound_data_buffer()
```
Comment (Lukasa) [inline test/test_basic_logic.py]: Also not needed.

**Comment by apolcyn** [comment]
removed the extra frames added to buffer, I think should be ready again

File: test/test_basic_logic.py [hunk 36]
```diff
@@ -109,10 +109,117 @@ def test_sending_data(self):
...
+        (ValueError, 256),
+        (TypeError, 'invalid'),
+        (TypeError, ''),
+        (TypeError, '10'),
+        (TypeError, {}),
+        (TypeError, ['1', '2', '3']),
+        (TypeError, []),
+        (TypeError, 1.5),
+        (TypeError, 1.0),
+        (TypeError, -1.0),
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: 👍

File: test/test_basic_logic.py [hunk 37]
```diff
@@ -1576,3 +1683,27 @@ def test_receiving_goaway_frame_with_unknown_error(self, frame_factory):
...
 
         assert not c.data_to_send()
+
+
+def sanity_check_data_frame(data_frame,
+                            expected_flow_controlled_length,
+                            expect_padded_flag,
+                            expected_data_frame_pad_length):
+        """
+        ```data_frame``` is a frame of type ```hyperframe.frame.DataFrame```,
```
Comment (alexwlchan) [inline test/test_basic_logic.py]: … triple quotes?

**Comment by Lukasa** [comment]
Yeah, squash and changelog would be good.

**Comment by alexwlchan** [comment]
> squash and changelog would be good

Done.

File: test/test_basic_logic.py (hunk 37 above)
Comment (apolcyn) [inline test/test_basic_logic.py]: ah woops I can still add to the branch, unless we should freeze it here?

**Comment by Lukasa** [comment]
Cool, looks great!

**Comment by Lukasa** [comment]
Thanks so much @apolcyn! Per our [one of the team policy](http://python-hyper.org/en/latest/one-of-the-team.html), you'll find an invite to the Python-hyper organisation in your inbox. Welcome on board!

**Comment by SethMichaelLarson** [comment]
:tada:

**Comment by apolcyn** [comment]
great thanks for the time in the review, this was really helpful!
